SYCL Kernel Support for Layer Normalization in oneDNN

AJAYRATNAM opened the pull request:
# Description
Introducing SYCL backend for the oneDNN Layer Normalization primitive. This contains support for Layernormalization functionality in both forward and backward propagations.

# Supported scope:
   Supported Data Types:
  Forward : f32, bf16, f16, s32, s8, u8
  Backward : f32, bf16, f16
  Attributes: Scales

## General
- [X] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [X] Have you formatted the code using clang-format?

## Testing Scope
    benchdnn: Passed

   Please refer the attached test output files.
    [test_outputs.zip](https://github.com/oneapi-src/oneDNN/files/10859429/test_outputs.zip)

densamoilov commented on src/gpu/sycl/ref_layer_normalizations.cpp:
```diff
@@ -0,0 +1,222 @@
+}
+
+status_t ref_layer_normalization_bwd_t::init(engine_t *engine) {
+    if (pd()->use_scale() || pd()->use_shift()) {
+        const auto kid
+                = ::sycl::get_kernel_id<layer_normalization_bwd_kernel_vec_t>();
+        CHECK(create_kernel(engine, kid, &kernel_));
+    }
+    const auto kid
+            = ::sycl::get_kernel_id<layer_normalization_bwd_kernel_vec_t2>();
+    CHECK(create_kernel(engine, kid, &kernel_));
+    return status::success;
```
Can you please explain logic behind this code? There is only 1 kernel named `kernel_` and in the case when we need a scale or shift we initialize `kernel_` with `layer_normalization_bwd_kernel_vec_t` but then we overwrite it by initializing it once again with `layer_normalization_bwd_kernel_vec_t2`.

densamoilov replied:
Any update on this? I see that the code hasn’t changed.

AJAYRATNAM replied:
Fixed kernel overwriting

densamoilov commented:
- I see a lot of unused variable warnings. Can you please fix them?
- I also hit an assert when creating a `sycl_md_t` from `src_md(1)` and `src_md(2)` because in the case when stats are not in src those src_mds are zero mds

densamoilov commented on src/gpu/sycl/ref_layer_normalizations.hpp:
```diff
@@ -0,0 +1,168 @@
+                    && (src_md(0)->format_desc.blocking.inner_nblks == 0)
+                    && (diff_dst_md(0)->format_desc.blocking.inner_nblks == 0)
+                    && utils::one_of(src_md(0)->data_type, f32, bf16, f16)
+                    && utils::one_of(diff_dst_md(0)->data_type, f32, bf16, f16)
+                    && utils::one_of(diff_src_md(0)->data_type, f32, bf16, f16)
```
We don't actually support backward for f16 data type, can you remove it?

densamoilov commented on src/gpu/nvidia/sycl_cuda_engine.cpp:
```diff
@@ -239,6 +239,10 @@ constexpr dnnl::impl::impl_list_item_t sycl_cuda_impl_list[] = {
         INSTANCE(sycl::ref_binary_t)
 
+        //layer norm
+        INSTANCE(sycl::ref_layer_normalization_fwd_t)
```
(random spot) Please add documentation for this primitive

saisrujana replied:
Yes. Added.

AJAYRATNAM commented:
We have accidentally discarded the commits while updating the fork. Reopening the PR.

AJAYRATNAM commented:
> * I see unused variable warnings. Can you please fix them?
> * I also hit an assert when creating a `sycl_md_t` from `src_md(1)` and `src_md(2)` because in the case when stats are not in src those src_mds are zero mds
> * There is another assert hit: src/common/primitive_exec_types.cpp:83: memory_t *dnnl::impl::exec_ctx_t::input(int) const: Assertion `ma.is_const' failed. This probably happens because there are arguments that can be both input and output depending on primitive parameters, so you need to adjust the code accordingly

Removed unused variables
Fixed assert hits

densamoilov commented on src/gpu/sycl/layer_normalizations_kernels.hpp:
```diff
@@ -0,0 +1,558 @@
+struct layer_normalization_fwd_kernel_vec_t1 {
+    static constexpr int vec_len = 8;
+
+    layer_normalization_fwd_kernel_vec_t1(
```
Please put number before `_t`

densamoilov commented on src/gpu/sycl/layer_normalizations_kernels.hpp:
```diff
@@ -0,0 +1,558 @@
+            sycl_in_memory_arg_t &scale, sycl_out_memory_arg_t &diff_scale,
+            sycl_out_memory_arg_t &diff_shift, sycl_in_memory_arg_t &stat,
+            sycl_in_memory_arg_t &var, sycl_in_memory_arg_t &diff_dst,
+            sycl_in_memory_arg_t &dst)
```
dst doesn't exist for backward, please remove md and memory for dst in backward pass.

densamoilov commented on src/gpu/sycl/ref_layer_normalizations.cpp:
```diff
@@ -0,0 +1,245 @@
+status_t ref_layer_normalization_fwd_t::pd_t::init_conf() {
+    conf_ = sycl_layer_normalization_conf_t();
+
+    conf_.var_md
```
I hit assert here when I run ` --lnorm --engine=gpu --dir=FWD_I --dt=f16:f16 --flags=CH --inplace=true 30x300`.
Both mds are zero mds for such case. You should also check that `desc_.prop_kind == prop_kind::forward_training`.

densamoilov commented on src/gpu/sycl/ref_layer_normalizations.cpp:
```diff
@@ -0,0 +1,245 @@
+    conf_.stat_md
+            = stats_are_src() ? sycl_md_t(src_md(1)) : sycl_md_t(dst_md(1));
+    conf_.dst_md = sycl_md_t(dst_md(0));
+    conf_.shift_off = conf_.use_ss && !has_zero_dim_memory()
```
`conf_.use_ss` is never initialized but used here.

densamoilov commented on src/gpu/sycl/ref_layer_normalizations.cpp:
```diff
@@ -0,0 +1,245 @@
+    conf_.stat_d = sycl_md_t(stat_md());
+    conf_.zero_dims = has_zero_dim_memory();
+    auto nelems_A = memory_desc_wrapper(src_md(0)).nelems();
+    conf_.diff_shift_off = conf_.use_ss && !conf_.zero_dims
```
conf_.use_ss is never initialized but used here.

densamoilov commented on src/gpu/sycl/ref_layer_normalizations.cpp:
```diff
@@ -0,0 +1,245 @@
+    }
+    const auto kid2
+            = ::sycl::get_kernel_id<layer_normalization_bwd_kernel_vec_t2>();
+    CHECK(create_kernel(engine, kid2, &kernel2_));
```
`CHECK(create_kernel(engine, kid2, &kernel2_));` -> `return create_kernel(engine, kid2, &kernel2_);`

densamoilov reviewed (changes requested):
I got a lot of failed cases when running test_lnorm_gpu:
```
[SRC][L0] = 15
[SRC][L1] exp:2.94891e+06 got:1.11411e+06 diff:2.52118e+06 rel_diff:0.854955
[SRC][L2] exp: 8002.13 got: 1730.43 diff: 7154.42 rel_diff:0.894064
[SRC][L8] exp:      64 got:       4 diff:      60 rel_diff:  0.9375
678:FAILED (errors:1 total:512) __REPRO: --lnorm --engine=gpu --dir=BWD_DW --flags=GCH --inplace=true 64x30x512
[SRC][L0] = 95324.1
[SRC][L1] exp:3.26532e+06 got:1.11411e+06 diff:2.87943e+06 rel_diff:0.881821
[SRC][L2] exp:  8042.7 got: 1730.43 diff: 7207.52 rel_diff:0.896158
[SRC][L8] exp: 72.3299 got:       4 diff: 68.3299 rel_diff:0.944698
679:FAILED (errors:1 total:512) __REPRO: --lnorm --engine=gpu --dir=BWD_DW --flags=CH --inplace=true 64x30x512
[SRC][L0] = 15
[SRC][L1] exp:2.94891e+06 got:1.11411e+06 diff:2.52118e+06 rel_diff:0.854955
[SRC][L2] exp: 8002.13 got: 1730.43 diff: 7154.42 rel_diff:0.894064
[SRC][L8] exp:      64 got:       4 diff:      60 rel_diff:  0.9375
680:FAILED (errors:1 total:512) __REPRO: --lnorm --engine=gpu --dir=BWD_DW --stat_tag=undef --flags=GCH --inplace=true 64x30x512
```

densamoilov reviewed (approved):
I went ahead and fixed remaining issues.

AJAYRATNAM commented:
> I went ahead and fixed remaining issues.

Thanks for Merging

saisrujana commented:
> I went ahead and fixed remaining issues.

Thanks a lot @densamoilov
